Fix access permission checks for distributed relations

metdos opened the pull request:
We ditched the access permission check with Custom Scan API changes. This PR brings it back.

onderkalaci commented on src/backend/distributed/planner/multi_planner.c:
```diff
@@ -295,6 +295,15 @@ CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query
 	/*
+	 * PostgreSQL checks access permissions in standard_ExecutorStart(), but
+	 * we do these checks here just before returning from planner. After this
+	 * point, we don't have access to our local tables in our plan. Therefore,
```
I think `local tables` is a bit confusing since we're using distributed relations in the next sentence. I'd rather use `original queries' range table list` or smth like that.

metdos replied:
Updated the comment.

metdos commented:
@onderkalaci refused to give a ship, so I moved him to reviewers :D

onderkalaci commented:
This looks good to me.

Let me share our conversation here on this:

* As far as we can see, the function should be called on the `standard_ExecutorStart()`
* After the custom scan node changes, we don't have access to the `standard_ExecutorStart()` (Postgres calls it for us!)
* In `standard_ExecutorStart()` , `ExecCheckRTPerms()` is already called. However, since the distributed plan is wrapped into a `customScan` node, the checks are done on the range tables of the `customScan`.
* Thus, we need to call that function just before the execution starts with the correct range tables. And, the appropriate place seems to be the end of `CreateDistributedPlan ()`

So, it's a :shipit: except for the very minor comment. But, let's give a chance to @anarazel and/or @jasonmp85 before merging in.

anarazel commented:
I don't think approach is sufficient. Wit prepared statements this is going to potentially happen in a different transaction than originally.

I'm also not following:
> In standard_ExecutorStart() , ExecCheckRTPerms() is already called. However, since the distributed plan is > wrapped into a customScan node, the checks are done on the range tables of the customScan.

The custom scan doesn't have a range table...  Isn't the fix here to just retain the original rtable in FinalizeRouterPlan()?

metdos commented:
> I don't think approach is sufficient. Wit prepared statements this is going to potentially happen in a different transaction than originally.

Darn, prepared statements.

I verified the problem here, we can extend regression tests.
https://github.com/citusdata/citus-enterprise/commit/09974b626c44245e246d4ed0e9d0b5bb1a2db5ec#diff-65502fee3d6eec89af3b80edfe2c121bR25

> The custom scan doesn't have a range table... Isn't the fix here to just retain the original rtable in FinalizeRouterPlan()?

@anarazel, are you proposing something like this?
```C
void
CitusSelectBeginScan(CustomScanState *node, EState *estate, int eflags)
{
	CitusScanState *scanState = (CitusScanState *) node;
	MultiPlan *multiPlan = scanState->multiPlan;
	Job *workerJob = multiPlan->workerJob;
	Query *jobQuery = workerJob->jobQuery;
	List *rangeTableList = jobQuery->rtable;

	ExecCheckRTPerms(rangeTableList, true);
}
```
If yes, currently range table entries are RTE_FUNCTION kind. I need to find a solution for this.

I couldn't look into this more today because of other code reviews etc. Monday, I will look into this more.

anarazel commented:
@metdos No, not really. In FinalizeRouterPlan() we currently do
```
	routerPlan = makeNode(PlannedStmt);
...
	routerPlan->rtable = list_make1(remoteScanRangeTableEntry);
```
so we replace the normal PlannedStmt with one that doesn't have the original rtable.  Which prevents ExecCheckRTPerms() from doing its job.   What I'm proposing is to just copy the rtable from `localPlan` (and thus set `customScanRangeTableIndex` to `list_length(localPlan->rtable)`).  That should take care of the whole issue, no?

metdos commented:
> What I'm proposing is to just copy the rtable from localPlan (and thus set customScanRangeTableIndex to list_length(localPlan->rtable)). That should take care of the whole issue, no?

This sounds good, but the same problem also exists for realtime-tasktracker executors. Therefore, I need to fix it also. OOTO currently, will take a look on Monday evening.

metdos commented:
I fixed the issue by adding original range table list to our finalized plan for both router and non-router executors. This has a side effect on the explain output. It adds table name (remote_scan) to column references which is fine I think.

I tested these changes on enterprise repo manually and with check-enterprise. I also added a test (https://github.com/citusdata/citus-enterprise/commit/e571b26a7ab4b7587bff90d86465baa9eb17a867#diff-980c54907a09b5a859437b45282df138R42) for prepared statements in permission checks test suite which we can merge after merging this PR.

Another possible fix is adding original range table to MultiPlan and running ExecCheckRTPerms() on this range table list explicitly at the beginning of the executions (CitusSelectBeginScan).

anarazel commented:
I think we should have a basic permission checking test case in this repo.

metdos commented:
Hey @jasonmp85, @anarazel, I added the regression test.

jasonmp85 commented:
Rebased this for myself/you.

jasonmp85 commented on src/test/regress/sql/multi_multiuser.sql:
```diff
@@ -0,0 +1,92 @@
+CREATE USER read_access;
+CREATE USER no_access;
+
+GRANT ALL ON TABLE test TO full_access;
```
Could probably have had slightly less noisy test output by doing this before distributing the table.

jasonmp85 commented on src/test/regress/sql/multi_multiuser.sql:
```diff
@@ -0,0 +1,92 @@
+GRANT SELECT ON TABLE test TO read_access;
+
+\c - - - :worker_1_port
+CREATE USER full_access;
```
Is this strictly necessary? The checks take place on the coordinator, no? I agree these statements on the worker make the test more realistic, but they're probably overkill to ensure permissions checks are happening.

sumedhpathak commented:
👍 Thanks Jason /Metin/Andres!

jasonmp85 commented:
Amusingly, it appears this new test file was not run against Enterprise, where it fails. Fixing.
